src/imp/mbedtls.rs

```diff
@@ -0,0 +1,575 @@
+        f.debug_struct("Identity")
+            .field(
+                "certificates",
+                &self
+                    .certificates
+                    .iter()
+                    .map(|cert| cert.as_der().to_vec())
+                    .collect::<Vec<_>>(),
+            )
+            .field(
+                "key_name",
+                &self.key.name().map(String::from).map_err(Error::Tls),
+            )
+            .finish()
+    }
+}
+
+#[derive(Clone)]
+pub struct Certificate(MbedtlsBox<MbedtlsCert>);
+
+impl Certificate {
+    pub fn from_der(buf: &[u8]) -> Result<Certificate, Error> {
+        let cert = MbedtlsCert::from_der(buf).map_err(Error::Tls)?;
+        Ok(Certificate(cert))
+    }
+
+    pub fn from_pem(buf: &[u8]) -> Result<Certificate, Error> {
+        let cert = MbedtlsCert::from_pem(&buf.null_terminated()).map_err(Error::Tls)?;
+        Ok(Certificate(cert))
+    }
+
+    pub fn to_der(&self) -> Result<Vec<u8>, Error> {
+        let der = self.0.as_der().to_vec();
+        Ok(der)
+    }
+}
+
+impl Debug for Certificate {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("Certificate")
+            .field(&self.0.as_der())
+            .finish()
+    }
+}
+
+pub struct TlsStream<S: Stream> {
+    ctx: Arc<Mutex<Context>>,
+    role: Endpoint,
+    identity: Option<Identity>,
+    _stream: PhantomData<S>,
+}
+
+impl<S: Stream> TlsStream<S> {
+    pub fn get_ref(&self) -> impl Deref<Target = S> + '_ {
+        // Is this unsound?
+        // map requires FnOnce(&mut T) -> &mut U, so I use io_mut()
+        // there. Return type is Deref, not DerefMut, so does it break
+        // the only single mutable reference rule?
+        // But this shouldn't be possible without unsafe
+        MutexGuard::map(self.ctx.lock(), |ctx| {
+            ctx.io_mut()
+                .expect("Not connected")
+                .downcast_mut::<S>()
+                // unwrap is safe as this is the exact stream passed to context
+                .unwrap()
+        })
+    }
+
+    pub fn get_mut(&mut self) -> impl DerefMut<Target = S> + '_ {
+        MutexGuard::map(self.ctx.lock(), |ctx| {
+            ctx.io_mut()
+                .expect("Not connected")
+                .downcast_mut::<S>()
+                // unwrap is safe as this is the exact stream passed to context
+                .unwrap()
+        })
+    }
+
+    pub fn buffered_read_size(&self) -> Result<usize, Error> {
+        Ok(self.ctx.lock().bytes_available())
+    }
+
+    #[cfg(feature = "alpn")]
+    pub fn negotiated_alpn(&self) -> Result<Option<Vec<u8>>, Error> {
+        use std::ffi::CStr;
+
+        let ctx_lock = self.ctx.lock();
+        let raw_ctx: *const mbedtls_sys::ssl_context = (&*ctx_lock).into();
+        let alpn = unsafe { CStr::from_ptr((*raw_ctx).alpn_chosen).to_bytes().to_vec() };
+        if alpn.is_empty() {
+            Ok(Some(alpn))
+        } else {
+            Ok(None)
+        }
```

Comment:
the condition should probably be negated, but need to check why the tests pass then :detective:

Reply:
fixed and added them to ci